src/Orchard.Web/Modules/Orchard.Fields/Drivers/BooleanFieldDriver.cs

```diff
@@ -35,7 +35,7 @@ public class BooleanFieldDriver : ContentFieldDriver<BooleanField> {
 
         protected override DriverResult Editor(ContentPart part, BooleanField field, dynamic shapeHelper) {
             // if the content item is new, assign the default value
-            if(!part.HasDraft() && !part.HasPublished()) {
+            if(!field.Value.HasValue) {
```

Comment:
You're right, by injecting an `IHttpContextAccessor` and using `_httpContextAccessor.Current()`. Then, one rule is to also use this mvc extension `if (!httpContext.IsBackgroundContext())` when it doesn't make sense to do something, e.g in a background task. But i think that the `Editor()` method will never be executed in the background (unlike the `Display()` method). Anyway, in the background an empty (but not null) route data would be returned.

So, not tested but i propose something like the following:
Note: maybe good to create a route extension `GetActionName()` as already done for `GetAreaName()`.

            ...
            var routeData = _httpContextAccessor.Current().Request.RequestContext.RouteData;

            object action;
            if (routeData.Values.TryGetValue("action", out action)) {
                var actionString = action as string;

                // if the content item is new, assign the default value
                if (!String.IsNullOrEmpty(actionString) && actionString.Equals("create", StringComparison.OrdinalIgnoreCase)) {
                    var settings = field.PartFieldDefinition.Settings.GetModel<BooleanFieldSettings>();
                    field.Value = settings.DefaultValue;
                }
            }
            ...